### PR TITLE
coap: ensure DTLS buffer is at least 200 bytes

### PR DESCRIPTION
**Pull request by benpicco**
### Contribution description

The initial DTLS handshake can be larger than a block in CoAP block-wise.
#19892 reduced the DTLS buffer to a CoAP block size + 36 bytes.
This has proven to be too little, breaking DTLS clients.

Bump `DTLS_MAX_BUF` to be at least 200 bytes as this has proven to work well in the past.

### Testing procedure

First, create the tap interfaces by running `sudo dist/tools/tapsetup/tapsetup`.

In one terminal, run the GCoAP DTSL server:

`make -C examples/gcoap_dtls PORT=tap1 all term`
```
2024-02-26 20:18:43,184 # main(): This is RIOT! (Version: 2024.04-devel-327-gb8d58-DTLS_MAX_BUF-fix)
2024-02-26 20:18:43,184 # gcoap example app
2024-02-26 20:18:43,185 # All up, running the shell now

2024-02-26 20:22:23,060 # > ifconfig
2024-02-26 20:22:23,060 # Iface  7  HWaddr: 5A:1A:98:23:2D:6C 
2024-02-26 20:22:23,061 #           L2-PDU:1500  MTU:1500  HL:64  Source address length: 6
2024-02-26 20:22:23,061 #           Link type: wired
2024-02-26 20:22:23,062 #           inet6 addr: fe80::581a:98ff:fe23:2d6c  scope: link  VAL
2024-02-26 20:22:23,063 #           inet6 group: ff02::1
2024-02-26 20:22:23,063 #           inet6 group: ff02::1:ff23:2d6c
```

In another terminal, run the nanoCoAP DTLS client:

`make -C tests/net/nanocoap_cli PORT=tap0 all term`
```
2024-02-26 20:23:47,051 # main(): This is RIOT! (Version: 2024.04-devel-326-gf04871)
2024-02-26 20:23:47,052 # nanocoap test app
2024-02-26 20:23:47,052 # All up, running the shell now

2024-02-26 20:23:52,901 # > ncget coaps://[fe80::581a:98ff:fe23:2d6c]/.well-known/core -
2024-02-26 20:23:52,905 # </cli/stats>;ct=0;rt="count";obs,</riot/board>
```

In `master` this would yield

```
2024-02-26 20:25:18,617 # main(): This is RIOT! (Version: 2024.04-devel-326-gf04871)
2024-02-26 20:25:18,618 # nanocoap test app
2024-02-26 20:25:18,618 # All up, running the shell now

2024-02-26 20:25:19,605 # > ncget coaps://[fe80::581a:98ff:fe23:2d6c]/.well-known/core -
2024-02-26 20:25:19,606 # cannot send ClientHello
2024-02-26 20:25:20,606 # cannot send ClientHello
2024-02-26 20:25:22,607 # cannot send ClientHello
```

### Issues/PRs references

follow-up to #19892

### CODE REVIEW

File: sys/net/application_layer/gcoap/Makefile.include [hunk 1]
```diff
@@ -3,5 +3,7 @@ ifeq (2, $(words $(filter ipv4 ipv6, $(USEMODULE))))
                          possible to use a dual stack setup$(COLOR_RESET)" 1>&2)
 endif
 
+max = $(subst xx,x,$(join $1,$2))
```
[the remark this answers is not in the document]
Comment (benpicco) [reply sys/net/application_layer/gcoap/Makefile.include:7]: join will sort the list, subst selects the 2nd element - I added a comment.

File: sys/net/application_layer/nanocoap/Makefile.include [hunk 2]
```diff
@@ -1,4 +1,8 @@
 ifneq (,$(filter nanocoap_dtls,$(USEMODULE)))
+  # join will sort the list, subst selects the 2nd element
+  max = $(subst xx,x,$(join $1,$2))
   CONFIG_NANOCOAP_BLOCKSIZE_DEFAULT := $(or $(CONFIG_NANOCOAP_BLOCKSIZE_DEFAULT),2)
-  DTLS_MAX_BUF ?= ((1 << ($(CONFIG_NANOCOAP_BLOCKSIZE_DEFAULT) + 3)) + 36)
+  # the initial DTLS handshake may exceed the block size
+  NANOCOAP_BLOCK_FRAME := $((1 << ($(CONFIG_NANOCOAP_BLOCKSIZE_DEFAULT) + 3)) + 36
```
Comment (fabian18) [inline sys/net/application_layer/nanocoap/Makefile.include:6]: I saw this in my editor. Here is a `)` missing

Comment (fabian18) [inline sys/net/application_layer/nanocoap/Makefile.include:6]: Does this kind of calculation work for you?

```makefile

  NANOCOAP_BLOCK_FRAME := $((1 << ($(CONFIG_NANOCOAP_BLOCKSIZE_DEFAULT) + 3)) + 36)
  $(info NANOCOAP_BLOCK_FRAME=$(NANOCOAP_BLOCK_FRAME))
  DTLS_MAX_BUF ?= $(call max,$(NANOCOAP_BLOCK_FRAME),200)
  $(info DTLS_MAX_BUF=$(DTLS_MAX_BUF))
```

prints

```
fabian.huessler@ml-pa.loc@MLPA-NB119:~/RIOT$ USEMODULE+="gnrc_icmpv6_echo nanocoap_dtls" BOARD=same54-xpro make -C tests/net/nanocoap_cli flash 
make: Entering directory '/home/fabian.huessler@ml-pa.loc/RIOT/tests/net/nanocoap_cli'
NANOCOAP_BLOCK_FRAME=
DTLS_MAX_BUF=200
```

And this:
```makefile
  NANOCOAP_BLOCK_FRAME := $(shell echo $$(( ((1 << ($(CONFIG_NANOCOAP_BLOCKSIZE_DEFAULT) + 3)) + 36) )))
  $(info NANOCOAP_BLOCK_FRAME=$(NANOCOAP_BLOCK_FRAME))
  DTLS_MAX_BUF ?= $(call max,$(NANOCOAP_BLOCK_FRAME),200)
  $(info DTLS_MAX_BUF=$(DTLS_MAX_BUF))
```

prints:

```
fabian.huessler@ml-pa.loc@MLPA-NB119:~/RIOT$ USEMODULE+="gnrc_icmpv6_echo nanocoap_dtls" BOARD=same54-xpro make -C tests/net/nanocoap_cli flash 
make: Entering directory '/home/fabian.huessler@ml-pa.loc/RIOT/tests/net/nanocoap_cli'
NANOCOAP_BLOCK_FRAME=68
DTLS_MAX_BUF=68200
```

**Comment by Teufelchen1** [comment]
*ding dong it's release' o'clock * Ping @benpicco

File: sys/net/application_layer/gcoap/Makefile.include (hunk 1 above)
[the remark this answers is not in the document]
Comment (fabian18) [reply sys/net/application_layer/gcoap/Makefile.include:7]: There is a really cool article about how to do math in a Makefile. It needs a bit more hacking than just `subst` and `join`. Maybe RIOT could just use the provided Makefile.
https://www.cmcrossroads.com/article/learning-gnu-make-functions-arithmetic

**Comment by benpicco** [comment]
I now use a `shell` call to calculate `DTLS_MAX_BUF` which gives much more sane values.

**Comment by benpicco** [comment]
@Teufelchen1 can this still get into the release? It fixes a regression.

**Review by Teufelchen1 — APPROVED** [review]

**Pull request merged**